docker/docker_client.go

```diff
@@ -513,6 +542,20 @@ func (c *dockerClient) makeRequestToResolvedURL(ctx context.Context, method stri
 	for {
 		res, err := c.makeRequestToResolvedURLOnce(ctx, method, url, headers, stream, streamLen, auth, extraScope)
 		attempts++
+
+		// By default we use pre-defined scopes per operation.  In certain
+		// cases, this can fail when our authentication is insufficient, then we
+		// might be getting an error back with a Www-Authenticate Header
+		// indicating an insufficient scope
+		// Check for that and update the client challenges to retry after
+		// requesting a new token
+		if retry, newScope := needsRetryWithUpdatedScope(err, res); retry && stream == nil && auth != noAuth {
+			logrus.Debug("Detected insufficient_scope error, will retry request with updated scope")
+			// Note: This retry ignores extraScope. That’s, strictly speaking, incorrect, but we don’t currently
+			// expect the insufficient_scope errors to happen for those callers. If that changes, we can add support
+			// for more than one extra scope.
+			res, err = c.makeRequestToResolvedURLOnce(ctx, method, url, headers, stream, streamLen, auth, newScope)
```

Comment:
@mtrmac Would it make sense to overwrite or merge `extraScope` with `newScope` here?

Reply:
Yes, I think that would be preferable — that way we do one extra request instead of N extra requests, against a server which we _know_ is overloaded.

---

Absolutely non-blocking: Longer-term, if we ever append to `extraScope` instead of replacing it entirely, we might need to worry about the server rejecting a scope every time, asking for more and more different variants on every retry.

So, maybe it would be ideal check+overwrite+immediately retry _only_ on the first iteration (`attempts == 1`); and let later retries only try once, with the updated scopes.

Reply:
I have moved the whole retry logic into an if statement so that we only do this on the first iteration and then overwrite the `extraScope`. However, I have not added any merging or anything like that, as this is just a single struct and no array, so in hindsight I don't know how I would merge it.

Reply:
Thanks! Perfect.

> However, I have not added any merging or anything like that, as this is just a single struct and no array, so in hindsight I don't know how I would merge it.

Conceptually, https://github.com/distribution/distribution/blob/main/docs/spec/auth/token.md#query-parameters suggests the client can ask for a token that gives access to multiple scopes. (But it also only says that about scopes that come from `WWW-Authenticate`, not scopes that the client hard-codes.)

You’re right that the code to actually do that doesn’t currently exist.